feat(sim): add X-Plane SITL airframes

alireza787b opened the pull request:
### Summary

Adds community-supported X-Plane SITL support using the external [`px4xplane`](https://github.com/alireza787b/px4xplane) bridge plugin.

This PR includes:
- X-Plane SITL airframe defaults for Cessna 172, TB2, EHang 184, Alia 250, and Quad Tailsitter.
- `make px4_sitl_default xplane_*` launch targets for those airframes.
- PX4 docs for installing the bridge plugin, selecting a target, and configuring the simulator host for local, WSL, Docker, or remote X-Plane setups.
- A community simulator index entry for X-Plane.

### Scope

The PR is intentionally limited to PX4-side airframes, target wiring, and documentation. The bridge plugin, X-Plane aircraft package, and release packaging remain in the external `px4xplane` repository.

### Validation

Local checks run on the cleaned branch:
- `git diff --cached --check` before commit
- Prettier check for the touched docs
- X-Plane airframe metadata generation for the five added airframes
- `make px4_sitl_default`
- Verified the generated Ninja graph contains the five X-Plane targets

### Notes

The branch was rebuilt from current `main` and squashed to one commit to resolve the previous conflicts. Airframe tuning is frozen for this PR unless reviewers request changes.

dagar commented on src/modules/simulation/simulator_mavlink/CMakeLists.txt:
```diff
@@ -71,3 +71,28 @@ add_custom_target(none_iris
+
+
+# xplane_c172 (legacy compatibility launch helper)
+add_custom_target(xplane_c172
```
maybe move these to a dedicated file like the others (eg sitl_targets_jsbsim.cmake, etc).

alireza787b replied:
Hello again
I just wanted to have an update on this pull request. should I do anything else before it merges? Since I want to create youtube tutorials on this topic I prefer not to instruct people to use my forked repo. I am waiting until this PR merge then create tutorial based on this.
thanks in advance

ryanjAA commented:
Really awesome!

ryanjAA commented:
are there non-windows plugin compiling instructions? or binaries (for Mac)
thanks! looking forward to trying this

alireza787b commented:
> are there non-windows plugin compiling instructions? or binaries (for Mac)
> thanks! looking forward to trying this

Well I am working on that ... I will add those instructions soon as well ... I have very limited experience building Cpp in macos...

ryanjAA commented:
What about in GitHub workflow?

alireza787b commented:
> What about in GitHub workflow?

I have no experience in that honestly.. but working on that with help of someone...

alireza787b commented:
Hello again 
I just wanted to have an update on this pull request. should I do anything else before it merges? Since I want to create youtube tutorials on this topic and I prefer not to instruct people use my forked repo. I am waiting until this PR merge then create tutorial based onthis . 
thanks in advance

ryanjAA commented:
Really looking forward to this! Any update on Mac?

alireza787b commented:
@dagar 

Hello,

I have marked this pull request as ready for review again. It includes the new airframes for the PX4 SITL simulation to interface with X-Plane, specifically the Cessna 172, Ehang 184, and Alia 250 and organized as you requested in last comments.

I am planning to create a YouTube tutorial on this feature, and it would be ideal to have it merged into the official repository to provide clear instructions to users. Could you please review the changes and let me know if any further modifications are required?

Thank you for your support!

Best regards,

hamishwillee commented on docs/en/sim_xplane/README.md:
```diff
@@ -0,0 +1,214 @@
+# X-Plane SITL Simulation
+
+X-Plane is a comprehensive flight simulator that can be used with PX4 for Software-In-The-Loop (SITL) simulation. This integration allows testing PX4 with highly realistic flight dynamics across multiple vehicle types.
```
```suggestion
:::warning
This simulator is [community supported and maintained](../simulation/community_supported_simulators.md).
It may or may not work with current versions of PX4 and may be removed in future releases.

See [Toolchain Installation](../dev_setup/dev_env.md) for information about the environments and tools supported by the core development team.
:::

X-Plane is a comprehensive flight simulator that can be used with PX4 for Software-In-The-Loop (SITL) simulation.
This integration allows testing PX4 with highly realistic flight dynamics across multiple vehicle types.
```

hamishwillee commented:
@alireza787b YOu just undid all the changes I made to your docs with the force push.

hamishwillee reviewed (commented):
Thanks @alireza787b -  I've given it a quick lookover and added a few comments. We need a technical reviewer.

@mahima-yoga Is this something you could look at, or recommend someone else?

mahima-yoga reviewed (commented):
Thanks for such well structured airframe files :smile:. I didn't have time yet to go through everything

alireza787b commented on ROMFS/px4fmu_common/init.d-posix/airframes/5001_xplane_cessna172:
```diff
@@ -0,0 +1,338 @@
+# Manual Control
+param set-default MPC_MAN_Y_MAX 20.0
+param set-default MPC_JERK_AUTO 1.0
+param set-default MPC_JERK_MAX 3.0
```
done... thanks... missed that

alireza787b commented on ROMFS/px4fmu_common/init.d-posix/airframes/5001_xplane_cessna172:
```diff
@@ -0,0 +1,338 @@
+param set-default FW_P_LIM_MIN -15.0     # Max pitch down
+param set-default FW_P_TC 0.45
+param set-default FW_PN_R_SLEW_MAX 50.0
+
```
Only parameters differing from PX4 defaults are included in configs (keeping files clean per your feedback on structure). Default
values (FW_YR_FF=0.3, FW_YR_D=0.0, FW_Y_RMAX=50 for Cessna/TB2) are omitted.
I will give each a further fine-tune while creating the demo video, and if I find better tuning values will commit again

mahima-yoga replied:
> Only parameters differing from PX4 defaults are included in configs

That's true, just making sure it was intentional

hamishwillee reviewed (dismissed):
Docs look good to me. Approving, but let's wait on your final changes to merge

hamishwillee commented:
You should check the [Checks / build (check_newlines) (pull_request)](https://github.com/PX4/PX4-Autopilot/actions/runs/19780376702/job/56679695360?pr=22493)
 fail and please let us know when you're ready for another look over

hamishwillee commented:
@mahima-yoga Can you confirm what needs to happen with this?

alireza787b commented:
Thanks @hamishwillee, and sorry for the slow response. I had limited connectivity during the regional internet disruption, but I am back on this now.

From my side, the remaining work is to rebase/check against current main, resolve the stale CI failures, and confirm whether the standard_vtol SITL failure is caused by this PR or unrelated base behavior. After that I will ask for another docs/simulation review.

@mahima-yoga / @StefanoColli, if there are any remaining technical blockers for the airframe configs, please point me to them and I will address them.

hamishwillee commented:
FYI @mahima-yoga is on holiday. Might take a little bit to respond.

mahima-yoga reviewed (dismissed):
From an airframe perspective it looks good to me! 

Please either clean up the commits before merging, or squash merge

hamishwillee commented:
Thanks for review @mahima-yoga ! @alireza787b Looks like we're good to go once the conflict fixed and commits are squashed

hamishwillee commented:
@alireza787b As above "Looks like we're good to go once the conflict fixed and commits are squashed". Can you do that?

Also, is the tuning finished? @mahima-yoga reviewed this but if you change things too much we'll have to get another review.

farhangnaderi commented:
@alireza787b Thanks. Whenever ready and conflicts are resolved as Hamish mentioned we can go for another review.

alireza787b commented:
Thanks @hamishwillee and @farhangnaderi. Quick update from my side, because I know I have been quiet longer than expected.

I did not want to just squash the old branch and ask for another review before checking the whole stack again. Over the last round I went back through the X-Plane side and the PX4 side with new logs:

- rebased/synced against current PX4 main and fixed stale params / current SITL startup defaults
- re-tested and tuned the five airframes again: Alia, Ehang, QuadTailsitter, Cessna 172, and TB2
- checked startup/sensor timing, low-FPS / pause behavior, land detection, and the noisy startup logs found during testing
- updated the px4xplane-side config workflow, config validation/editor UX, camera config support, and test docs/cards outside this PX4 branch
- cleaned the PX4 airframe parameter files so the defaults are intentional and easier to review

I also cleaned the PR branch now: it is squashed to one commit, and the diff no longer carries the EKF2 core change or the temporary logger cleanup. This PR is now scoped to X-Plane SITL airframes, target wiring, the small simulator startup handling needed for X-Plane defaults, and docs.

During the final Alia transition tests against current main I found a separate EKF-GSF emergency yaw-reset edge case. In short: GNSS velocity aiding and magnetic heading were healthy, but a transient GNSS position rejection could still allow an EKF-GSF yaw reset just after VTOL handoff. In my logs that reset yaw to a bad GSF estimate and then produced a false compass fault.

I opened that as a focused EKF PR here:
https://github.com/PX4/PX4-Autopilot/pull/27533

That PR is passing CI now.

Question for reviewers: do you prefer that I continue finalizing this X-Plane PR independently while the EKF PR is reviewed separately, or should I treat this PR as temporarily waiting on the EKF PR for final VTOL transition validation?

alireza787b committed:
feat(sim): add X-Plane SITL airframes

Add X-Plane SITL build targets and airframe startup defaults for Cessna 172, TB2, Ehang 184, Alia 250, and QuadTailsitter.

Document the community-supported X-Plane workflow, including TCP connection behavior and WSL/remote host setup.

Keep the branch scoped to X-Plane simulator integration; estimator changes are handled separately.

hamishwillee committed:
docs(doc): Alpabetic ordering

hamishwillee commented on docs/en/sim_xplane/index.md:
```diff
@@ -0,0 +1,87 @@
+## Installation
+
+1. Install the [PX4 development environment](../dev_setup/dev_env.md).
+2. Install X-Plane 11 or X-Plane 12.
```
IF there is a great installation link, would be good to provide.

hamishwillee commented on docs/en/sim_xplane/index.md:
```diff
@@ -0,0 +1,87 @@
+1. Install the [PX4 development environment](../dev_setup/dev_env.md).
+2. Install X-Plane 11 or X-Plane 12.
+3. Download a px4xplane release from the [px4xplane releases page](https://github.com/alireza787b/px4xplane/releases).
+4. Install the bridge plugin into:
```
What bridge plugin are you refering to?

hamishwillee commented on docs/en/sim_xplane/index.md:
```diff
@@ -0,0 +1,87 @@
+
+1. Install the [PX4 development environment](../dev_setup/dev_env.md).
+2. Install X-Plane 11 or X-Plane 12.
+3. Download a px4xplane release from the [px4xplane releases page](https://github.com/alireza787b/px4xplane/releases).
```
Which one? Basically I'm looking at these points and I think you're assuming that PX4 developers know something about X-Plane. Maybe they do, but a little hand holding and assuming they know nothing might be better.

hamishwillee commented on docs/en/sim_xplane/index.md:
```diff
@@ -0,0 +1,87 @@
+See [Toolchain Installation](../dev_setup/dev_env.md) for information about the environments and tools supported by the core development team.
+:::
+
+[X-Plane](https://www.x-plane.com/) can be used with PX4 for [Software-In-The-Loop (SITL)](../simulation/index.md#sitl-simulation-environment) simulation through the community-supported [px4xplane](https://github.com/alireza787b/px4xplane) bridge plugin.
```
Does it work on any OS/all OS supported by PX4? Can I run it in WSL2?

hamishwillee commented on docs/en/sim_xplane/index.md:
````diff
@@ -0,0 +1,87 @@
+   X-Plane/Resources/plugins/px4xplane
+   ```
+
+5. Install the matching X-Plane aircraft and bridge configuration from the release package.
````
Is this the package I just downloaded from " Download a px4xplane release " step? Is that a double click? The less users have to think the more likely they are to try it out.

hamishwillee commented:
@mahima-yoga @farhangnaderi Can one of you please go through this quickly again as it has had lots of change since your last check, as outlined in https://github.com/PX4/PX4-Autopilot/pull/22493#issuecomment-4589317993.

alireza787b committed:
docs(sim): clarify X-Plane setup workflow

alireza787b commented:
Thanks @hamishwillee, good point. The page was assuming too much X-Plane/px4xplane context.

I pushed a small docs-only update in `23d6fa4`:

- clarified that the px4xplane bridge is the X-Plane plugin and what it does between PX4 and X-Plane
- made the release install flow explicit: download the OS-specific archive, extract it, copy the plugin folder, and copy the matching aircraft files
- clarified the Windows + WSL2 workflow vs same-machine Linux/macOS usage
- adjusted the community simulator row so it says these are the preconfigured PX4 targets, while custom aircraft need a matching PX4 airframe and px4xplane bridge config

I kept the PX4 docs intentionally short and linked out to the px4xplane docs for the plugin layout/custom-airframe details.

I also left your docs ordering commit in place. If you prefer the branch squashed again after this review pass, I can do that.

alireza787b committed:
docs(sim): clarify X-Plane setup and custom models

alireza787b commented:
One more small docs-only follow-up in `271dc46` after rereading the setup page from a new user point of view. clarified that `PX4_SIM_HOSTNAME` is set on the PX4/SITL side and should point to the machine running X-Plane + the plugin, including WSL2/Docker/remote-host cases. I also added the TCP 4560 firewall note, linked the official X-Plane Plane Maker/dataref references, and adjusted the custom-model wording so it is clear the shipped targets are examples, while new models still need a matching PX4 control path and writable dataref mappings.\n\nNo code or parameter changes in this push.

alireza787b commented:
Final update from my side before the next review pass: px4xplane v4.0.0 is now published with Windows, Linux, and macOS plugin packages: https://github.com/alireza787b/px4xplane/releases/tag/v4.0.0\n- the px4xplane README/docs were cleaned up for the current workflow, including WSL2/Docker host-IP setup, config editor docs, custom-airframe guidance, and the video archive/future playlist link\n- this PX4 PR is green in CI and still scoped to the X-Plane SITL targets/docs/startup wiring\n\nThe separate EKF-GSF yaw-reset issue found during final Alia VTOL transition testing is still isolated in #27533. That PR is also green in CI. Until that lands, a fast VTOL transition on current PX4 main can still hit the false compass/yaw-reset warning in that specific edge case, but I am keeping that fix separate from this X-Plane integration PR.

hamishwillee commented:
Docs look better thanks.

> I also left your docs ordering commit in place. If you prefer the branch squashed again after this review pass, I can do that.

Fine by me. It will all be squashed on merge.

mahima-yoga commented on ROMFS/px4fmu_common/init.d-posix/airframes/5001_xplane_cessna172:
```diff
@@ -0,0 +1,312 @@
+
+param set-default MPC_MAN_Y_MAX 20.0
+param set-default MPC_JERK_AUTO 1.0
+param set-default MPC_JERK_MAX 3.0
```
These multicopter parameters can be removed.

mahima-yoga commented on ROMFS/px4fmu_common/init.d-posix/airframes/5002_xplane_tb2:
```diff
@@ -0,0 +1,301 @@
+param set-default FW_AIRSPD_STALL 20.0
+param set-default FW_AIRSPD_FLP_SC 0.85
+param set-default FW_USE_AIRSPD 1
+param set-default ASPD_DO_CHECKS 1
```
How come all the other checks are disabled?

mahima-yoga reviewed (changes requested):
There are a lot of parameters that are set to their default value - and this adds a lot of unnecessary bloat to the airframe files. Usually we only want to set the the non-default parameters. 

Some examples I found were: 

- Many EKF2 parameters 
- SYS_HAS parameters 
- FW_USE_AIRSPD 
- FW_THR_IDLE
- FW_THR_MAX
- RWTO_NUDGE
- LNDFW_TRIG_TIME

Would you mind removing the default parameters?

alireza787b committed:
fix(sim): trim X-Plane airframe defaults

alireza787b commented:
Thanks @mahima-yoga. I pushed `e1634873` for this.

I went through all five X-Plane airframe files and removed the params that were only restating inherited PX4 defaults. I checked this against the generated parameter defaults plus the relevant `rc.fw_defaults`, `rc.mc_defaults`, and `rc.vtol_defaults`, so the tuned values stay in the airframe files and the default duplicates are gone.

A couple of notes on intentional behavior:

- `ASPD_DO_CHECKS` is no longer overridden. Cessna, TB2, Alia, and QuadTailsitter now inherit the PX4 default checks instead of forcing only the missing-data check.
- `SYS_HAS_NUM_ASPD=0` is still kept for the X-Plane airspeed-enabled frames. The pitot data comes from the simulator/HIL stream, not a physical airspeed driver that should block prearm. The aircraft still use modeled airspeed for control/fallback where configured.
- The Cessna leftover `MPC_*` lines are removed. They were not part of the fixed-wing control path.

Local checks I ran before pushing:

- duplicate-default scan over all five X-Plane airframes: zero redundant inherited defaults remaining
- `git diff --check`
- `bash -n` on all five airframe scripts
- `make px4_sitl_default`
- launched `xplane_cessna172`, `xplane_tb2`, `xplane_ehang184`, `xplane_alia250`, and `xplane_qtailsitter` up to the simulator TCP wait; no parameter-not-found or shell errors

CI is running again now.

alireza787b committed:
fix(sim): restore X-Plane validation margins

alireza787b committed:
fix(sim): correct X-Plane accel calibration

Leave simulated accelerometer Z offsets at zero for the packaged X-Plane airframes where the HIL vertical specific force already matches gravity. Recent logs showed the negative Z offsets made parked aircraft report about 0.7-0.8 m/s^2 too little gravity, which led EKF2 to accumulate vertical velocity and emit preflight warnings on the ground.\n\nCarry over the latest Cessna runway/TECS tuning from validation: a higher takeoff airspeed margin, slower roll slew, modest takeoff flaps, and smoother TECS damping.

alireza787b committed:
fix(sim): smooth Alia front transition

Tune the standard VTOL transition parameters based on Alia validation logs. The pusher was saturated through transition while pitch was not established before the handoff to fixed-wing, causing altitude loss immediately after entering FW mode.\n\nUse PX4's standard VTOL tuning parameters instead of changing controller logic: reduce transition pusher throttle and slew, keep MC lift active to a higher blend airspeed, complete the handoff slightly earlier, and increase the pitch offset/feed-forward used during and after front transition.

alireza787b committed:
fix(sim): refine X-Plane transition margins

Retune the Alia standard VTOL front transition after validation showed the previous low pusher setting could not reliably reach transition speed before the watchdog. Keep enough pusher thrust for an on-time transition, start pitch/MC-FW blending earlier, and retain a stronger FW pitch response through handoff.\n\nAlso raise the X-Plane HIL GPS velocity noise consistently across the simulator airframes. The simulator publishes zero parked GPS velocity, but frame-rate-limited HIL startup/ground operation can still produce small EKF vertical velocity innovations; the updated covariance better matches the simulator sensor contract without disabling estimator checks.

alireza787b committed:
fix(sim): tune X-Plane longitudinal energy

Use validation logs from Alia and Cessna to reduce post-transition/post-takeoff speed chasing. Alia entered fixed-wing around transition speed but then dove while TECS chased an overly high trim speed; Cessna similarly sank after climbout when flaps retracted and normal FW guidance chased cruise speed with aggressive bank.\n\nTune airframe defaults through PX4's standard speed, TECS, pitch-limit, transition, and NPFG parameters. This keeps the change scoped to aircraft configuration and avoids simulator-specific controller logic.

alireza787b committed:
fix(sim): improve Alia FW energy handoff

mahima-yoga reviewed (dismissed):
Thanks for all the detailed descriptions, comments and tests!

alireza787b committed:
fix(sim): tune X-Plane Alia and Cessna energy

alireza787b committed:
fix(sim): refine Alia and Cessna climbout energy

alireza787b commented:
> Thanks for all the detailed descriptions, comments and tests!

Thanks, I pushed one final validation update after retesting Alia and C172 against the current branch.

  This last push is still scoped to X-Plane airframe validation defaults: Alia front-transition pitch offset, and C172 climbout/flap/energy margins after runway takeoff. CI is
  running again now.

  During final VTOL validation I also isolated two PX4-side edge cases separately:
  - #27533 for the EKF-GSF yaw-reset/compass warning case seen in fast VTOL transition.
  - #27601 for Standard VTOL front-transition setpoint handoff. That one is still draft while I finish validation and get maintainer feedback.

  I am doing one last local end-to-end pass with those guard PRs applied locally before saying this is ready to merge. Please hold merge until I confirm the final validation
  result.

alireza787b committed:
fix(sim): use default accel bias process noise

alireza787b committed:
fix(sim): refine Alia elevator and EKF defaults

alireza787b committed:
fix(sim): use neutral X-Plane accel calibration

alireza787b commented:
Final update from my side. The local end-to-end validation pass is done now, and I no longer need this PR held for more X-Plane-side changes.

What I rechecked after the last upstream sync:

- all five X-Plane targets: Alia 250, Cessna 172, TB2, Ehang 184, and QuadTailsitter
- clean SITL startup with fresh saved parameters / reset-config flow
- WSL2 host-IP workflow and the px4xplane launcher sync path
- X-Plane docs from a new-user point of view: bridge/plugin meaning, release install, WSL2/Docker/remote-host IP setup, TCP 4560, custom aircraft / Plane Maker / dataref mapping links
- airframe params after the duplicate-default cleanup requested by @mahima-yoga

This PR is still scoped to the X-Plane SITL integration: targets, airframes, startup wiring, and docs. While validating against current upstream PX4 I also found a few PX4-side controller/estimator edge cases, so I kept them out of this branch and split them into separate PRs:

- #27533 - EKF-GSF emergency yaw reset guard for a fast VTOL transition edge case
- #27601 - Standard VTOL front-transition pusher-thrust / pitch-setpoint handoff
- #27669 - Tailsitter fixed-wing attitude-frame consistency for quaternion attitude error
- #27670 - Fixed-wing TECS reset when the local altitude reference changes

Those separate PRs explain the delay in the final validation pass after rebasing/syncing to current PX4. I used them locally in the validation stack where needed, but they are intentionally not bundled into this X-Plane SITL PR.

From my side #22493 is ready for final review/merge. If there is any remaining blocker in this branch, please point me to it and I will fix it directly.

hamishwillee reviewed (dismissed):
Thanks @alireza787b - @mahima-yoga Can you merge this if you are happy? The docs seem sufficient :-)

mahima-yoga reviewed (dismissed):
So sorry that I let this one slip. Looks good to merge, if we find a small issue/mis-configuration we can always fix it later but I think it's time to bring this in now 😄 

Could you do a rebase to resolve the conflicts? And then let's merge it!

alireza787b commented:
Thanks @mahima-yoga. I rebased this on current `main` and resolved the docs conflict by keeping both the upstream RotorPy entry and the X-Plane entry.\n\nLocal checks after the rebase:\n- `make px4_sitl_default`\n- `git diff --check`\n- conflict-marker scan over the touched docs/startup files\n\nThe branch is mergeable again now. CI restarted after the force-push.

hamishwillee reviewed (approved):
Thanks @alireza787b @mahima-yoga 

Merging based on @mahima-yoga review and request for rebase (but then merge).
